File: codeql-custom-queries-java/android/avoid-media-resource-leak.java
class MediaRecorder {
    void start() {}
    void stop() {}
    void release() {}
}

class MediaPlayer {
    void start() {}
    void stop() {}
    void release() {}
}

class NoncompliantMediaLeak {

    // 🚫 Noncompliant - MediaRecorder field never released
    MediaRecorder leakedRecorder = new MediaRecorder(); // $ Alert

    // 🚫 Noncompliant - MediaPlayer field never released
    MediaPlayer leakedPlayer = new MediaPlayer(); // $ Alert

    void recordAudio() {
        // 🚫 Noncompliant - local MediaRecorder never released
        MediaRecorder recorder = new MediaRecorder(); // $ Alert
        recorder.start();
        recorder.stop();
        // missing recorder.release()
    }

    void playAudio() {
        // 🚫 Noncompliant - local MediaPlayer never released
        MediaPlayer player = new MediaPlayer(); // $ Alert
        player.start();
        player.stop();
        // missing player.release()
    }

    void recordAndPlay() {
        // 🚫 Noncompliant - both never released
        MediaRecorder r = new MediaRecorder(); // $ Alert
        MediaPlayer p = new MediaPlayer();     // $ Alert
        r.start();
        p.start();
    }
}

class CompliantMediaUsage {

    void recordAudio() {
        // ✅ Compliant - release() called
        MediaRecorder recorder = new MediaRecorder(); // OK
        recorder.start();
        recorder.stop();
        recorder.release();
    }

    void playAudio() {
        // ✅ Compliant - release() called
        MediaPlayer player = new MediaPlayer(); // OK
        player.start();
        player.stop();
        player.release();
    }

    void recordAndPlay() {
        // ✅ Compliant - both released
        MediaRecorder r = new MediaRecorder(); // OK
        MediaPlayer p = new MediaPlayer();     // OK
        r.start();
        p.start();
        r.release();
        p.release();
    }
}

File: codeql-custom-queries-java/android/avoid-media-resource-leak.kt
// Stubs
class MediaRecorder {
    fun release() {}
    fun start() {}
    fun stop() {}
}

class MediaPlayer {
    fun release() {}
    fun start() {}
    fun stop() {}
}

// 🚫 Noncompliant - no release() called
class NoncompliantMediaLeak {

    fun recordAudio() {
        val recorder = MediaRecorder() // $ Alert
        recorder.start()
        recorder.stop()
        // missing recorder.release()
    }

    fun playAudio() {
        val player = MediaPlayer() // $ Alert
        player.start()
        player.stop()
        // missing player.release()
    }

    val leakedRecorder = MediaRecorder() // $ Alert
    val leakedPlayer = MediaPlayer()     // $ Alert
}

// ✅ Compliant - release() is called
class CompliantMediaUsage {

    fun recordAudio() {
        val recorder = MediaRecorder() // OK
        recorder.start()
        recorder.stop()
        recorder.release()
    }

    fun playAudio() {
        val player = MediaPlayer() // OK
        player.start()
        player.stop()
        player.release()
    }
}